File: swift-bridge/Sources/MapKitBridge/MapItem.swift
import CoreLocation
import Foundation
import MapKit

struct MKRPlacemarkPayload: Codable {
    var coordinate: MKRCoordinatePayload
    var countryCode: String?
    var title: String?
}

struct MKRMapItemPayload: Codable {
    var identifier: String?
    var alternateIdentifiers: [String]
    var name: String?
    var phoneNumber: String?
    var url: String?
    var timeZoneIdentifier: String?
    var pointOfInterestCategory: String?
    var isCurrentLocation: Bool
    var placemark: MKRPlacemarkPayload?
    var location: MKRCoordinatePayload?
    var address: MKRAddressPayload?
    var addressRepresentations: MKRAddressRepresentationsPayload?
}

func mkrBuildPlacemark(_ payload: MKRPlacemarkPayload) -> MKPlacemark {
    MKPlacemark(coordinate: mkrCoordinate(from: payload.coordinate))
}

func mkrEncodePlacemark(_ placemark: MKPlacemark) -> MKRPlacemarkPayload {
    MKRPlacemarkPayload(
        coordinate: mkrEncodeCoordinate(placemark.coordinate),
        countryCode: placemark.countryCode,
        title: placemark.title
    )
}

func mkrBuildMapItem(_ payload: MKRMapItemPayload) throws -> MKMapItem {
    if payload.isCurrentLocation && payload.placemark == nil && payload.location == nil {
        return MKMapItem.forCurrentLocation()
    }

    let item: MKMapItem
    if #available(macOS 26.0, *), let location = payload.location {
        let address = try mkrBuildAddress(payload.address)
        item = MKMapItem(
            location: CLLocation(
                latitude: location.latitude,
                longitude: location.longitude
            ),
            address: address
        )
    } else if let placemark = payload.placemark {
        item = MKMapItem(placemark: mkrBuildPlacemark(placemark))
    } else {
        throw NSError(
            domain: "mapkit-rs",
            code: -1,
            userInfo: [NSLocalizedDescriptionKey: "MKMapItem payload is missing a location or placemark"]
        )
    }

    item.name = payload.name
    item.phoneNumber = payload.phoneNumber
    item.url = payload.url.flatMap(URL.init(string:))
    item.timeZone = payload.timeZoneIdentifier.flatMap(TimeZone.init(identifier:))
    item.pointOfInterestCategory = payload.pointOfInterestCategory.map(mkrPointOfInterestCategory(from:))
    return item
}

func mkrEncodeMapItem(_ item: MKMapItem) -> MKRMapItemPayload {
    let identifier: String?
    let alternateIdentifiers: [String]
    if #available(macOS 15.0, *) {
        identifier = item.identifier?.rawValue
        alternateIdentifiers = item.alternateIdentifiers.map(\.rawValue)
    } else {
        identifier = nil
        alternateIdentifiers = []
    }

    let location: MKRCoordinatePayload?
    if #available(macOS 26.0, *) {
        location = mkrEncodeCoordinate(item.location.coordinate)
    } else if item.isCurrentLocation {
        location = nil
    } else {
        location = mkrEncodeCoordinate(item.placemark.coordinate)
    }

    let placemark = item.isCurrentLocation ? nil : mkrEncodePlacemark(item.placemark)
    let address: MKRAddressPayload?
    let addressRepresentations: MKRAddressRepresentationsPayload?
    if #available(macOS 26.0, *) {
        address = mkrEncodeAddress(item.address)
        addressRepresentations = mkrEncodeAddressRepresentations(item.addressRepresentations)
    } else {
        address = nil
        addressRepresentations = nil
    }

    return MKRMapItemPayload(
        identifier: identifier,
        alternateIdentifiers: alternateIdentifiers,
        name: item.name,
        phoneNumber: item.phoneNumber,
        url: item.url?.absoluteString,
        timeZoneIdentifier: item.timeZone?.identifier,
        pointOfInterestCategory: item.pointOfInterestCategory?.rawValue,
        isCurrentLocation: item.isCurrentLocation,
        placemark: placemark,
        location: location,
        address: address,
        addressRepresentations: addressRepresentations
    )
}

struct MKRMapItemRequestStatePayload: Codable {
    var mapItemIdentifier: String?
    var cancelled: Bool
    var loading: Bool
}

private enum MKRMapItemStringConstantKind: Int32 {
    case launchOptionsCameraKey = 0
    case launchOptionsDirectionsModeCycling = 1
    case launchOptionsDirectionsModeDefault = 2
    case launchOptionsDirectionsModeDriving = 3
    case launchOptionsDirectionsModeKey = 4
    case launchOptionsDirectionsModeTransit = 5
    case launchOptionsDirectionsModeWalking = 6
    case launchOptionsMapCenterKey = 7
    case launchOptionsMapSpanKey = 8
    case launchOptionsMapTypeKey = 9
    case launchOptionsShowsTrafficKey = 10
    case mapItemTypeIdentifier = 11
}

@_cdecl("mk_map_item_string_constant")
public func mk_map_item_string_constant(
    _ kind: Int32,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    do {
        guard let kind = MKRMapItemStringConstantKind(rawValue: kind) else {
            throw NSError(
                domain: "mapkit-rs",
                code: -1,
                userInfo: [NSLocalizedDescriptionKey: "unknown MKMapItem string constant kind"]
            )
        }

        let value: String
        switch kind {
        case .launchOptionsCameraKey:
            value = MKLaunchOptionsCameraKey
        case .launchOptionsDirectionsModeCycling:
            value = MKLaunchOptionsDirectionsModeCycling
        case .launchOptionsDirectionsModeDefault:
            value = MKLaunchOptionsDirectionsModeDefault
        case .launchOptionsDirectionsModeDriving:
            value = MKLaunchOptionsDirectionsModeDriving
        case .launchOptionsDirectionsModeKey:
            value = MKLaunchOptionsDirectionsModeKey
        case .launchOptionsDirectionsModeTransit:
            value = MKLaunchOptionsDirectionsModeTransit
        case .launchOptionsDirectionsModeWalking:
            value = MKLaunchOptionsDirectionsModeWalking
        case .launchOptionsMapCenterKey:
            value = MKLaunchOptionsMapCenterKey
        case .launchOptionsMapSpanKey:
            value = MKLaunchOptionsMapSpanKey
        case .launchOptionsMapTypeKey:
            value = MKLaunchOptionsMapTypeKey
        case .launchOptionsShowsTrafficKey:
            value = MKLaunchOptionsShowsTrafficKey
        case .mapItemTypeIdentifier:
            value = MKMapItemTypeIdentifier
        }
        return mkrCString(value)
    } catch {
        mkrSetError(outError, error)
        return nil
    }
}

@_cdecl("mk_map_item_request_new")
public func mk_map_item_request_new(
    _ mapItemIdentifier: UnsafePointer<CChar>?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutableRawPointer? {
    guard #available(macOS 15.0, *) else {
        mkrSetMessageError(outError, message: "MKMapItemRequest requires macOS 15.0")
        return nil
    }
    guard let mapItemIdentifier else {
        mkrSetMessageError(outError, message: "missing MKMapItemIdentifier")
        return nil
    }

    let identifierString = String(cString: mapItemIdentifier)
    guard let identifier = MKMapItem.Identifier(rawValue: identifierString) else {
        mkrSetMessageError(
            outError,
            message: "invalid MKMapItemIdentifier raw value: \(identifierString)"
        )
        return nil
    }

    return mkrRetain(MKMapItemRequest(mapItemIdentifier: identifier))
}

@_cdecl("mk_map_item_request_state_json")
public func mk_map_item_request_state_json(
    _ request: UnsafeMutableRawPointer?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    guard #available(macOS 15.0, *) else {
        mkrSetMessageError(outError, message: "MKMapItemRequest requires macOS 15.0")
        return nil
    }
    guard let request else {
        mkrSetMessageError(outError, message: "missing MKMapItemRequest")
        return nil
    }

    do {
        let bridge = mkrBorrow(request, as: MKMapItemRequest.self)
        let payload = MKRMapItemRequestStatePayload(
            mapItemIdentifier: bridge.mapItemIdentifier?.rawValue,
            cancelled: bridge.isCancelled,
            loading: bridge.isLoading
        )
        return mkrCString(try mkrEncodeJSON(payload))
    } catch {
        mkrSetError(outError, error)
        return nil
    }
}

@_cdecl("mk_map_item_request_get_map_item_json")
public func mk_map_item_request_get_map_item_json(
    _ request: UnsafeMutableRawPointer?,
    _ outError: UnsafeMutablePointer<UnsafeMutablePointer<CChar>?>?
) -> UnsafeMutablePointer<CChar>? {
    guard #available(macOS 15.0, *) else {
        mkrSetMessageError(outError, message: "MKMapItemRequest requires macOS 15.0")
        return nil
    }
    guard let request else {
        mkrSetMessageError(outError, message: "missing MKMapItemRequest")
        return nil
    }

    let bridge = mkrBorrow(request, as: MKMapItemRequest.self)
    do {
        let mapItem: MKMapItem = try mkrAwaitOnMain { completion in
            bridge.getMapItem(completionHandler: { mapItem, error in
                if let mapItem {
                    completion(.success(mapItem))
                } else {
                    completion(.failure(error ?? NSError(
                        domain: "mapkit-rs",
                        code: -1,
                        userInfo: [NSLocalizedDescriptionKey: "MKMapItemRequest completed without a map item"]
                    )))
                }
            })
        }
        return mkrCString(try mkrEncodeJSON(mkrEncodeMapItem(mapItem)))
    } catch {
        mkrSetError(outError, error)
        return nil
    }
}

@_cdecl("mk_map_item_request_cancel")
public func mk_map_item_request_cancel(_ request: UnsafeMutableRawPointer?) {
    guard #available(macOS 15.0, *), let request else { return }
    let bridge = mkrBorrow(request, as: MKMapItemRequest.self)
    bridge.cancel()
}

@_cdecl("mk_map_item_request_release")
public func mk_map_item_request_release(_ request: UnsafeMutableRawPointer?) {
    guard let request else { return }
    mkrRelease(request)
}

File: src/map_item.rs
use core::ffi::c_void;
use std::ptr::{self, NonNull};

use serde::{Deserialize, Serialize};

use crate::address::{MKAddress, MKAddressRepresentations};
use crate::error::MapKitError;
use crate::ffi;
use crate::geometry::MKCoordinate;
use crate::private::{cstring_from_str, owned_handle, parse_json_ptr, take_string};

/// Wraps `MKPlacemark`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MKPlacemark {
    /// Wraps `MKPlacemark.coordinate`.
    pub coordinate: MKCoordinate,
    /// Wraps `MKPlacemark.countryCode`.
    pub country_code: Option<String>,
    /// Wraps `MKPlacemark.title`.
    pub title: Option<String>,
}

impl MKPlacemark {
    /// Creates a wrapper for `MKPlacemark`.
    pub const fn new(coordinate: MKCoordinate) -> Self {
        Self {
            coordinate,
            country_code: None,
            title: None,
        }
    }

    /// Wraps `MKPlacemark.countryCode`.
    pub fn with_country_code(mut self, country_code: impl Into<String>) -> Self {
        self.country_code = Some(country_code.into());
        self
    }

    /// Wraps `MKPlacemark.title`.
    pub fn with_title(mut self, title: impl Into<String>) -> Self {
        self.title = Some(title.into());
        self
    }
}

/// Wraps `MKMapItemIdentifier`.
#[derive(Debug, Clone, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(transparent)]
pub struct MKMapItemIdentifier(pub String);

impl MKMapItemIdentifier {
    /// Creates a wrapper for `MKMapItemIdentifier`.
    pub fn new(raw_value: impl Into<String>) -> Self {
        Self(raw_value.into())
    }

    /// Wraps `MKMapItemIdentifier.rawValue`.
    pub fn raw_value(&self) -> &str {
        &self.0
    }
}

impl AsRef<str> for MKMapItemIdentifier {
    fn as_ref(&self) -> &str {
        self.raw_value()
    }
}

/// Wraps `MKMapItem`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MKMapItem {
    /// Wraps `MKMapItem.identifier`.
    pub identifier: Option<String>,
    /// Wraps `MKMapItem.alternateIdentifiers`.
    #[serde(default)]
    pub alternate_identifiers: Vec<String>,
    /// Wraps `MKMapItem.name`.
    pub name: Option<String>,
    /// Wraps `MKMapItem.phoneNumber`.
    pub phone_number: Option<String>,
    /// Wraps `MKMapItem.url`.
    pub url: Option<String>,
    /// Wraps `MKMapItem.timeZoneIdentifier`.
    pub time_zone_identifier: Option<String>,
    /// Wraps `MKMapItem.pointOfInterestCategory`.
    pub point_of_interest_category: Option<String>,
    /// Wraps `MKMapItem.isCurrentLocation`.
    pub is_current_location: bool,
    /// Wraps `MKMapItem.placemark`.
    pub placemark: Option<MKPlacemark>,
    /// Wraps `MKMapItem.location`.
    pub location: Option<MKCoordinate>,
    /// Wraps `MKMapItem.address`.
    pub address: Option<MKAddress>,
    /// Wraps `MKMapItem.addressRepresentations`.
    pub address_representations: Option<MKAddressRepresentations>,
}

impl MKMapItem {
    /// Creates a wrapper for `MKMapItem`.
    pub fn new(placemark: MKPlacemark) -> Self {
        let coordinate = placemark.coordinate;
        Self {
            identifier: None,
            alternate_identifiers: Vec::new(),
            name: None,
            phone_number: None,
            url: None,
            time_zone_identifier: None,
            point_of_interest_category: None,
            is_current_location: false,
            placemark: Some(placemark),
            location: Some(coordinate),
            address: None,
            address_representations: None,
        }
    }

    /// Wraps `MKMapItem.fromLocation`.
    pub fn from_location(location: MKCoordinate, address: Option<MKAddress>) -> Self {
        Self {
            identifier: None,
            alternate_identifiers: Vec::new(),
            name: None,
            phone_number: None,
            url: None,
            time_zone_identifier: None,
            point_of_interest_category: None,
            is_current_location: false,
            placemark: None,
            location: Some(location),
            address,
            address_representations: None,
        }
    }

    /// Wraps `MKMapItem.currentLocation`.
    pub fn current_location() -> Self {
        Self {
            identifier: None,
            alternate_identifiers: Vec::new(),
            name: None,
            phone_number: None,
            url: None,
            time_zone_identifier: None,
            point_of_interest_category: None,
            is_current_location: true,
            placemark: None,
            location: None,
            address: None,
            address_representations: None,
        }
    }

    /// Wraps `MKMapItem.name`.
    pub fn with_name(mut self, name: impl Into<String>) -> Self {
        self.name = Some(name.into());
        self
    }

    /// Wraps `MKMapItem.phoneNumber`.
    pub fn with_phone_number(mut self, phone_number: impl Into<String>) -> Self {
        self.phone_number = Some(phone_number.into());
        self
    }

    /// Wraps `MKMapItem.url`.
    pub fn with_url(mut self, url: impl Into<String>) -> Self {
        self.url = Some(url.into());
        self
    }

    /// Wraps `MKMapItem.timeZoneIdentifier`.
    pub fn with_time_zone_identifier(mut self, time_zone_identifier: impl Into<String>) -> Self {
        self.time_zone_identifier = Some(time_zone_identifier.into());
        self
    }

    /// Wraps `MKMapItem.pointOfInterestCategory`.
    pub fn with_point_of_interest_category(
        mut self,
        point_of_interest_category: impl Into<String>,
    ) -> Self {
        self.point_of_interest_category = Some(point_of_interest_category.into());
        self
    }

    /// Wraps `MKMapItem.address`.
    pub fn with_address(mut self, address: MKAddress) -> Self {
        self.address = Some(address);
        self
    }

    /// Wraps `MKMapItem.coordinate`.
    pub fn coordinate(&self) -> Option<MKCoordinate> {
        self.location.or_else(|| {
            self.placemark
                .as_ref()
                .map(|placemark| placemark.coordinate)
        })
    }

    /// Wraps `MKMapItem.identifierValue`.
    pub fn identifier_value(&self) -> Option<MKMapItemIdentifier> {
        self.identifier.clone().map(MKMapItemIdentifier)
    }

    /// Wraps `MKMapItem.alternateIdentifierValues`.
    pub fn alternate_identifier_values(&self) -> Vec<MKMapItemIdentifier> {
        self.alternate_identifiers
            .iter()
            .cloned()
            .map(MKMapItemIdentifier)
            .collect()
    }

    /// Wraps `MKMapItem.launchOptionsCameraKey`.
    pub fn launch_options_camera_key() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsCameraKey,
            "MKLaunchOptionsCameraKey",
        )
    }

    /// Wraps `MKMapItem.launchOptionsDirectionsModeCycling`.
    pub fn launch_options_directions_mode_cycling() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsDirectionsModeCycling,
            "MKLaunchOptionsDirectionsModeCycling",
        )
    }

    /// Wraps `MKMapItem.launchOptionsDirectionsModeDefault`.
    pub fn launch_options_directions_mode_default() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsDirectionsModeDefault,
            "MKLaunchOptionsDirectionsModeDefault",
        )
    }

    /// Wraps `MKMapItem.launchOptionsDirectionsModeDriving`.
    pub fn launch_options_directions_mode_driving() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsDirectionsModeDriving,
            "MKLaunchOptionsDirectionsModeDriving",
        )
    }

    /// Wraps `MKMapItem.launchOptionsDirectionsModeKey`.
    pub fn launch_options_directions_mode_key() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsDirectionsModeKey,
            "MKLaunchOptionsDirectionsModeKey",
        )
    }

    /// Wraps `MKMapItem.launchOptionsDirectionsModeTransit`.
    pub fn launch_options_directions_mode_transit() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsDirectionsModeTransit,
            "MKLaunchOptionsDirectionsModeTransit",
        )
    }

    /// Wraps `MKMapItem.launchOptionsDirectionsModeWalking`.
    pub fn launch_options_directions_mode_walking() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsDirectionsModeWalking,
            "MKLaunchOptionsDirectionsModeWalking",
        )
    }

    /// Wraps `MKMapItem.launchOptionsMapCenterKey`.
    pub fn launch_options_map_center_key() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsMapCenterKey,
            "MKLaunchOptionsMapCenterKey",
        )
    }

    /// Wraps `MKMapItem.launchOptionsMapSpanKey`.
    pub fn launch_options_map_span_key() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsMapSpanKey,
            "MKLaunchOptionsMapSpanKey",
        )
    }

    /// Wraps `MKMapItem.launchOptionsMapTypeKey`.
    pub fn launch_options_map_type_key() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsMapTypeKey,
            "MKLaunchOptionsMapTypeKey",
        )
    }

    /// Wraps `MKMapItem.launchOptionsShowsTrafficKey`.
    pub fn launch_options_shows_traffic_key() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::LaunchOptionsShowsTrafficKey,
            "MKLaunchOptionsShowsTrafficKey",
        )
    }

    /// Wraps `MKMapItem.typeIdentifier`.
    pub fn type_identifier() -> Result<String, MapKitError> {
        map_item_string_constant(
            MKMapItemStringConstantKind::MapItemTypeIdentifier,
            "MKMapItemTypeIdentifier",
        )
    }
}

#[derive(Debug, Clone, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
struct MKMapItemRequestState {
    map_item_identifier: Option<String>,
    cancelled: bool,
    loading: bool,
}

/// Wraps `MKMapItemRequest`.
#[derive(Debug)]
pub struct MKMapItemRequest {
    raw: NonNull<c_void>,
}

impl MKMapItemRequest {
    /// Creates a wrapper for `MKMapItemRequest`.
    pub fn new(map_item_identifier: &MKMapItemIdentifier) -> Result<Self, MapKitError> {
        let identifier = cstring_from_str(map_item_identifier.raw_value(), "MKMapItemIdentifier")?;
        let mut error = ptr::null_mut();
        let raw = unsafe { ffi::mk_map_item_request_new(identifier.as_ptr(), &mut error) };
        let raw = owned_handle(raw, error, "failed to create MKMapItemRequest")?;
        Ok(Self { raw })
    }

    fn state(&self) -> Result<MKMapItemRequestState, MapKitError> {
        let mut error = ptr::null_mut();
        let payload = unsafe { ffi::mk_map_item_request_state_json(self.raw.as_ptr(), &mut error) };
        if payload.is_null() {
            Err(unsafe {
                MapKitError::from_error_ptr(error, "failed to read MKMapItemRequest state")
            })
        } else {
            unsafe { parse_json_ptr(payload, "MKMapItemRequest state") }
        }
    }

    /// Wraps `MKMapItemRequest.mapItemIdentifier`.
    pub fn map_item_identifier(&self) -> Result<Option<MKMapItemIdentifier>, MapKitError> {
        Ok(self.state()?.map_item_identifier.map(MKMapItemIdentifier))
    }

    /// Wraps `MKMapItemRequest.isCancelled`.
    pub fn is_cancelled(&self) -> Result<bool, MapKitError> {
        Ok(self.state()?.cancelled)
    }

    /// Wraps `MKMapItemRequest.isLoading`.
    pub fn is_loading(&self) -> Result<bool, MapKitError> {
        Ok(self.state()?.loading)
    }

    /// Wraps `MKMapItemRequest.mapItem`.
    pub fn map_item(&self) -> Result<MKMapItem, MapKitError> {
        let mut error = ptr::null_mut();
        let payload =
            unsafe { ffi::mk_map_item_request_get_map_item_json(self.raw.as_ptr(), &mut error) };
        if payload.is_null() {
            Err(unsafe { MapKitError::from_error_ptr(error, "MKMapItemRequest getMapItem failed") })
        } else {
            unsafe { parse_json_ptr(payload, "MKMapItem") }
        }
    }

    /// Wraps `MKMapItemRequest.cancel`.
    pub fn cancel(&self) {
        unsafe { ffi::mk_map_item_request_cancel(self.raw.as_ptr()) };
    }
}

impl Drop for MKMapItemRequest {
    fn drop(&mut self) {
        unsafe { ffi::mk_map_item_request_release(self.raw.as_ptr()) };
    }
}

#[derive(Clone, Copy)]
#[repr(i32)]
enum MKMapItemStringConstantKind {
    LaunchOptionsCameraKey = 0,
    LaunchOptionsDirectionsModeCycling = 1,
    LaunchOptionsDirectionsModeDefault = 2,
    LaunchOptionsDirectionsModeDriving = 3,
    LaunchOptionsDirectionsModeKey = 4,
    LaunchOptionsDirectionsModeTransit = 5,
    LaunchOptionsDirectionsModeWalking = 6,
    LaunchOptionsMapCenterKey = 7,
    LaunchOptionsMapSpanKey = 8,
    LaunchOptionsMapTypeKey = 9,
    LaunchOptionsShowsTrafficKey = 10,
    MapItemTypeIdentifier = 11,
}

fn map_item_string_constant(
    kind: MKMapItemStringConstantKind,
    context: &str,
) -> Result<String, MapKitError> {
    let mut error = ptr::null_mut();
    let value = unsafe { ffi::mk_map_item_string_constant(kind as i32, &mut error) };
    if value.is_null() {
        Err(unsafe { MapKitError::from_error_ptr(error, context) })
    } else {
        unsafe {
            take_string(value).ok_or_else(|| {
                MapKitError::OperationFailed(format!("missing string payload for {context}"))
            })
        }
    }
}
